daemon/cluster/executor/container/controller.go

```diff
@@ -153,10 +154,40 @@ func (r *controller) Wait(pctx context.Context) error {
 	ctx, cancel := context.WithCancel(pctx)
 	defer cancel()
 
-	err := r.adapter.wait(ctx)
-	if ctx.Err() != nil {
+	// error chan for wait
+	waitErrCh := make(chan error)
+	// error chan for health check
+	healthErrCh := make(chan error)
+
+	runner := func(f func(ctx context.Context) error, errCh chan<- error) {
+		err := f(ctx)
+		select {
+		case <-r.closed:
+		case <-ctx.Done():
+		case errCh <- err:
+		}
+	}
+
+	// run wait and checkHealth
+	go runner(r.adapter.wait, waitErrCh)
+	go runner(r.checkHealth, healthErrCh)
```

Comment:
@tonistiigi, 2 go routines are used, instead of one as you suggested. Because using one go routine I got some tricky synchronization problems.

I tested health check on 2 restart conditions, `any` and `on_failure`. It works correctly and exited unhealthy container gets exit code 137.

Is it ok to keep 2 go routine here?